rust/src/websocket/detect.rs

```diff
@@ -0,0 +1,134 @@
+/* Copyright (C) 2023 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::websocket::WebSocketTransaction;
+use crate::detect::uint::{detect_parse_uint, DetectUintData, DetectUintMode};
+use crate::detect::Enum;
+use crate::websocket::parser::WebSocketOpcode;
+
+use nom7::branch::alt;
+use nom7::bytes::complete::{is_a, tag};
+use nom7::combinator::{opt, value};
+use nom7::multi::many1;
+use nom7::IResult;
+
+use std::ffi::CStr;
+
+#[no_mangle]
+pub unsafe extern "C" fn SCWebSocketGetOpcode(tx: &mut WebSocketTransaction) -> u8 {
+    return tx.pdu.opcode;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCWebSocketGetFlags(tx: &mut WebSocketTransaction) -> u8 {
+    return tx.pdu.flags;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCWebSocketGetPayload(
+    tx: &WebSocketTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    *buffer = tx.pdu.payload.as_ptr();
+    *buffer_len = tx.pdu.payload.len() as u32;
+    return true;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCWebSocketGetMask(
+    tx: &mut WebSocketTransaction, value: *mut u32,
+) -> bool {
+    if let Some(xorkey) = tx.pdu.mask {
+        *value = xorkey;
+        return true;
+    }
+    return false;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCWebSocketParseOpcode(
+    ustr: *const std::os::raw::c_char,
+) -> *mut DetectUintData<u8> {
+    let ft_name: &CStr = CStr::from_ptr(ustr); //unsafe
+    if let Ok(s) = ft_name.to_str() {
+        if let Some(ctx) = WebSocketOpcode::to_detect_ctx(s) {
+            let boxed = Box::new(ctx);
+            return Box::into_raw(boxed) as *mut _;
+        }
+    }
+    return std::ptr::null_mut();
+}
```

Comment:
As we return `DetectUintData<u8>`, it would be cleaner if we parsed that I think, like:
```
	if let Ok(ctx) = DetectUintData::<u8>::from_str(s) {
	    let boxed = Box::new(ctx);
	    return Box::into_raw(boxed) as *mut _;
	}
```
would require this impl though:
```
impl<T: DetectIntType> FromStr for DetectUintData<T> {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
	if let Ok((_, v)) = detect_parse_uint::<T>(s) {
	    Ok(v)
	} else {
	    Err(())
	}
    }
}
```

Reply:
How does the patch wok for `websocket.opcode: 1` ? (numeric value instead of string value)

Reply:
> How does the patch wok for `websocket.opcode: 1` ? (numeric value instead of string value)

Oh, the tests must not test for that as S-V did pass.

I do have the trait there so you can do:
```
let ctx = DetectUintData::from_str(s)
```
as well.  So you would try one that the other.  It just moves the conversions to be a long side their datatype, rather than somewhere that seems a little random.

Reply:
> It just moves the conversions to be a long side their datatype, rather than somewhere that seems a little random.

Seems right, see https://github.com/OISF/suricata/pull/10110/commits/d5f1bf2f1196fad64de37e9ce922aebadc6c6a85 for next iteration